### PR TITLE
Use window.document for React-Native

### LINKED ISSUE #38286: @microsoft/signalr V6.0.0 is not working on react-native apps
I updated @microsoft/signalr library on my react native app (v0.66.2) to the new version (v6.0.0) and now i get this error:

"HubConnection failed to start successfully because of error 'ReferenceError: Can't find variable: document'."

**On previous version (v5.0.12) everything was working fine.**

is there a way to fix this on react-native?

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes #38286

### CODE REVIEW

File: src/SignalR/clients/ts/signalr/src/HubConnection.ts [hunk 1]
```diff
@@ -180,9 +180,9 @@ export class HubConnection {
             await this._startInternal();
 
             if (Platform.isBrowser) {
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/HubConnection.ts:182]: It looks like our `Platform.isBrowser` check is broken. It's currently `return typeof window === "object"` which is always `true` for React Native even outside the browser. I just tested this on my Android phone.

When running React Native inside the browser, you can still reference `document` directly, and I think we still do want to attach the "freeze" event listener in that case.

I think changing `Platform.isBrowser` to `return typeof document === "object"` might be the better fix.

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 2]
```diff
@@ -35,16 +35,20 @@ export class Arg {
...
+    // WebWorkers don't have a window object so the isBrowser check would fail
     public static get isWebWorker(): boolean {
         return typeof self === "object" && "importScripts" in self;
     }
 
+    // Node apps shouldn't have a window object, but WebWorkers don't either
+    // so we need to check for both WebWorker and window
     public static get isNode(): boolean {
-        return !this.isBrowser && !this.isWebWorker;
+        return !this.isBrowser && !this.isWebWorker && typeof window !== "object";
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/Utils.ts:51]:
```suggestion
        return !this.isBrowser && !this.isWebWorker && !this.isReactNative();
```

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 3]
```diff
@@ -35,16 +35,20 @@ export class Arg {
 
 /** @private */
 export class Platform {
+    // react-native has a window but no document so we should check both
     public static get isBrowser(): boolean {
-        return typeof window === "object";
+        return typeof window === "object" && !!window.document;
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/Utils.ts:40]: Nit:

```suggestion
        return typeof window === "object" && typeof window.document === "object";
```
```suggestion
        return typeof window === "object" && !!window.document;
```

File: src/SignalR/clients/ts/signalr/src/Utils.ts [hunk 4]
```diff
@@ -37,18 +37,23 @@ export class Arg {
...
     }
 
     // WebWorkers don't have a window object so the isBrowser check would fail
     public static get isWebWorker(): boolean {
         return typeof self === "object" && "importScripts" in self;
     }
 
+    // react-native has a window but no document
+    static get isReactNative(): boolean {
+        return typeof window === "object" && window.document === undefined;
```
Comment (halter73) [inline src/SignalR/clients/ts/signalr/src/Utils.ts:50]:
```suggestion
        return typeof window === "object" && typeof window.document === "undefined";
```

Why not this for consistency with all the other checks?

**Review by halter73 — APPROVED** [review]

**Comment by hikaaam** [comment]
this issue not fixed yet ?
```
 LOG  [2021-11-24T10:20:30.130Z] Debug: The HttpConnection connected successfully.
 LOG  [2021-11-24T10:20:30.132Z] Debug: Sending handshake request.
 INFO  [2021-11-24T10:20:30.136Z] Information: Using HubProtocol 'json'.
 LOG  [2021-11-24T10:20:30.390Z] Debug: Server handshake complete.
 LOG  [2021-11-24T10:20:30.393Z] Debug: HubConnection failed to start successfully because of error 'ReferenceError: Property 'document' doesn't exist'.
```

**Comment by BrennanConroy** [comment]
If you're using nightly bits it is, but if you're using 6.0 it isn't. Use the workaround at https://github.com/dotnet/aspnetcore/issues/38286#issuecomment-970580861

**Pull request merged**